### ISSUE TITLE
stb_image is it thread safe?

### ISSUE THREAD

**Issue by giordi91** [nothings/stb#309]
Hi, I am using stb_image to load texture for my game engine, all works nice and dandy but I am getting what I believe being race conditions inside stbi_load and I would like to know If that function is thread safe. What I am doing is really simple:

`auto data= stbi_load(path,&m_width, &m_height, 0, 0);`

This is at the start of my own Texture::load, that class is allocated on the heap right before calling the load, so no race condition can happen there ,the first line in load() is the stbi_load function you see.
From time to time, an assertion from stbi_load (or any nested func call), is being triggered.
The assertion is the following:

Assertion failed: z->size[b] == s, file E:\WORK_IN_PROGRESS\C\libs\Simple-OpenGL-Image-Library\src\stb_image_aug.c, line 1750

Which looking at the current version i have (to be honest the stb_image i am using came from SOIL, but i have bypassed it completely and called stb_image directly):

`__forceinline static int zhuffman_decode(zbuf *a, zhuffman *z)
{
   int b,s,k;
   if (a->num_bits < 16) fill_bits(a);
   b = z->fast[a->code_buffer & ZFAST_MASK];
   if (b < 0xffff) {
      s = z->size[b];
      a->code_buffer >>= s;
      a->num_bits -= s;
      return z->value[b];
   }

   // not resolved by fast table, so compute it the slow way
   // use jpeg approach, which requires MSbits at top
   k = bit_reverse(a->code_buffer, 16);
   for (s=ZFAST_BITS+1; ; ++s)
      if (k < z->maxcode[s])
         break;
   if (s == 16) return -1; // invalid code!
   // code size is s, so:
   b = (k >> (16-s)) - z->firstcode[s] + z->firstsymbol[s];
   assert(z->size[b] == s);
   a->code_buffer >>= s;
   a->num_bits -= s;
   return z->value[b];
}`

Now anyone know what could be triggering this? If I move just the stbi_func call inside a lock everything works perfectly, so I would be curious to know whats going on and how to prevent the assertion, having texture loading inside a lock is less than ideal and would rather avoid it.

**Comment by giordi91** [comment]
I updated to the latest and completely removed soil and looks like the problem is gone, still interested to know what might have caused it but for now all looks ok.

**Comment by glampert** [comment]
Hi,

I believe [this](https://github.com/nothings/stb/blob/master/stb_image.h#L3801) could be the source of your problem (from `stb_image.h:3801`):

```cpp
 // @TODO: should statically initialize these for optimal thread safety
static stbi_uc stbi__zdefault_length[288], stbi__zdefault_distance[32];
```

The library is *mostly* thread safe, but some of the loaders (not for all image formats) rely on some shared static data. The PNG loader from the link above is one such case.

The easiest fix, I think, is to search for the static variables in the file and mark them with a *thread local* qualifier (varies for each compiler, see for instance [`__declspec(thread)`](https://msdn.microsoft.com/en-us/library/9w1sdazb.aspx)).

If I recall, there are only a handful of static variables which could cause a race condition, so it is not a huge task (another case I remember is the `stbi__g_failure_reason`, which you might want to also make thread local to get proper error messages for each thread).

Hope this is of help, @giordi91.

I would also be interested in hearing @nothings thoughts on this. What would you think is the best way of going about those static vars, like the failure reason and those Zlib buffers? Marking them as thread local? Maybe introducing a new API with a "loader context" parameter perhaps? Is this something you think is worth fixing?

**Comment by nothings** [comment]
__declspec(thread) isn't portable to old compilers. Since those tables are always read-only, they just need to be safely initialized, which is probably easiest done by just embedding the initial values in the source code, or we could add an explicit init function (which only needs to be called if you're using it from muitiple threads).

for "failure_reason", I don't know of a good solution without changing the API, which I really don't want to do. maybe for that the time has come to figure out how to do thread locals portably on everything anyone cares about.

**Comment by lieff** [comment]
May be it`s better do implicit initialization
static const stbi_uc stbi__zdefault_length[288] = { ... };
under STB_IMAGE_IMPLEMENTATION, and
extern const stbi_uc stbi__zdefault_length[];
without STB_IMAGE_IMPLEMENTATION.

stbi__init_zdefaults loops can be unrolled by compiler and function may become larger than 288+32 bytes.
Another bonus with const qualifier global variable goes to .rdata section (which is protected) and can not be changed at runtime (without mmap play).

**Comment by giordi91** [comment]
I wish I could help more but I don't know exactly the nature of the data involved, but yeah if there is a wya to statically initialize it would be the best solution I guess. So just to clarify race condition is still there even in the latest version correct? The reason why I am not getting race conditions is just probably luck or different stuff happening different time giving me the luck of avoiding data race, but the possibility is still there. If I understand correctly that data is initialized when the first png is loaded? If so loading a dummy image from the main thread would patch the problem for the time being?

**Comment by nothings** [comment]
Maybe, loading one PNG before any threaded loading can happen could avoid it, but only if the PNG uses the default zlib tables, which an arbitrary PNG may not do (in fact probably won't do). So I'm not even sure if that path *is* the path you were running into.

But anyway, to address that case a simpler way is to just have that thread call "stbi__init_zdefaults()" in the file that defines STB_IMAGE_IMPLEMENTATION. Obviously this isn't part of the public API, but I can make sure I keep it around and with the same functionality when I fix it properly.

**Comment by nothings** [comment]
Reopening the issue since I really *should* fix this properly at some point.

**Comment by giordi91** [comment]
what would be the case of a png using default zlib tables? i can try to make a case where that happens just to stay on the safe side.

**Comment by nothings** [comment]
It's a property of the zlib compression spec. I don't know when the PNG compressor decides to use the default tables (if ever), and it depends on the PNG as well. I guess stb_image_write does.

**Comment by Shaderhacker** [comment]
Does this apply to the jpg files too?  I'm getting access violation crashes launching all available threads to load separate images into main memory.  What's interesting is that I only get it in Windows 10.  I've compiled a version for Linux and I can get all 16 threads chugging along fine loading these 4k images.  In Windows 10, it seems like I can only get a few threads through but as I increase the number of threads, it crashes.  

Crash site is here:

static int stbi__jpeg_decode_block_prog_dc(stbi__jpeg *j, short data[64], stbi__huffman *hdc, int b)

particularly this call:

**Comment by giordi91** [comment]
I was on win 10 too.

**Comment by DJLink** [comment]
I ran into this problem too, was making some tests on 3DS, and it's super fast to load with stb_image, but launching from threads leads to the problem already pointed. Wonder if anyone found an "elegant" solution

**Comment by nothings** [comment]
I looked over the JPG code and didn't see any threading issues, but I may have missed something. @DJLink, is your "too" referring to PNG or JPG?

**Comment by DJLink** [comment]
@nothings PNG. using stbi_load_from_memory
I can get it working without using threads perfectly, but as soon as I use them puff. Trying to find exactly what is causing it

**Comment by nothings** [comment]
As mentioned upthread, for PNG, just try calling stbi__init_zdefaults() before any thread does any decoding. This is still a bug, it's just a probably workaround for now.

**Comment by rygorous** [comment]
The issue in question was fixed in commit 351489803fbd1d5a9ad62d6bb0c86f2726072314.
